IfBuildProfile does not work on scheduler classes

I have the following class

```java
@IfBuildProfile(anyOf = {"dev", "prod", "RateLimitSchedulerProfile"})
public class RateLimitScheduler {

    @Scheduled(every = "1s")
    public void schedule() {
        System.out.println("hello " + System.currentTimeMillis());
    }
}
```

And the following test

```java
public class RateLimitSchedulerProfile implements QuarkusTestProfile {
    @Override
    public String getConfigProfile() {
        return "test,RateLimitSchedulerProfile";
    }
}

@QuarkusTest
@TestProfile(RateLimitSchedulerProfile.class)
class RateLimitSchedulerTest {

    @Inject
    RateLimitScheduler rateLimitScheduler;

    @Test
    void test() {
        System.out.println("RateLimitSchedulerTest:");
        rateLimitScheduler.schedule();
    }

}
```

This test fails with 

```text
[ERROR] Errors: 
[ERROR]   HelloResourceTest.testHelloEndpoint » Runtime java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: jakarta.enterprise.inject.spi.DeploymentException: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type gr.RateLimitScheduler and qualifiers [@Default]
	- injection target: gr.RateLimitSchedulerTest#rateLimitScheduler
	- declared on CLASS bean [types=[gr.RateLimitSchedulerTest, java.lang.Object], qualifiers=[@Default, @Any], target=gr.RateLimitSchedulerTest]
	The following classes match by type, but have been skipped during discovery:
	- gr.RateLimitScheduler was annotated with @Vetoed


	at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1576)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:338)
	at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:178)
	at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:492)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)
Caused by: jakarta.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type gr.RateLimitScheduler and qualifiers [@Default]
	- injection target: gr.RateLimitSchedulerTest#rateLimitScheduler
	- declared on CLASS bean [types=[gr.RateLimitSchedulerTest, java.lang.Object], qualifiers=[@Default, @Any], target=gr.RateLimitSchedulerTest]
	The following classes match by type, but have been skipped during discovery:
	- gr.RateLimitScheduler was annotated with @Vetoed


	at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:547)
	at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:697)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:323)
	... 12 more
```

In order to "fix" it i would have to use the `%test.quarkus.scheduler.enabled=false` and delete the `@IfBuildProfile` from RateLimitSchedulerProfile class.

Note: The test which fails is not the RateLimitSchedulerTest but the HelloResourceTest. If i disable the HelloResourceTest it works.

[Reproducer](https://github.com/gbourant/quarkus-scheduler-bug)

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TestsAsBeansProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TestsAsBeansProcessor.java
@@ -6,12 +6,15 @@ import java.util.List;
 import java.util.Optional;
 import java.util.Set;
 
+import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
 import org.jboss.jandex.AnnotationTarget.Kind;
+import org.jboss.jandex.AnnotationTransformation;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.ClassInfo.NestingType;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
+import org.jboss.logging.Logger;
 
 import io.quarkus.arc.processor.Annotations;
 import io.quarkus.arc.processor.AnnotationsTransformer;
@@ -25,6 +28,16 @@ import io.quarkus.deployment.builditem.TestClassBeanBuildItem;
 import io.quarkus.deployment.builditem.TestProfileBuildItem;
 
 public class TestsAsBeansProcessor {
+
+    private static final Logger LOG = Logger.getLogger(TestsAsBeansProcessor.class);
+
+    private static final DotName QUARKUS_TEST_PROFILE = DotName.createSimple("io.quarkus.test.junit.QuarkusTestProfile");
+    private static final DotName TEST_PROFILE = DotName.createSimple("io.quarkus.test.junit.TestProfile");
+    private static final DotName QUARKUS_TEST = DotName.createSimple("io.quarkus.test.junit.QuarkusTest");
+    private static final DotName QUARKUS_INTEGRATION_TEST = DotName
+            .createSimple("io.quarkus.test.junit.QuarkusIntegrationTest");
+    private static final DotName QUARKUS_MAIN_TEST = DotName.createSimple("io.quarkus.test.junit.main.QuarkusMainTest");
+    private static final DotName NESTED = DotName.createSimple("org.junit.jupiter.api.Nested");
 
     @BuildStep
     public void testAnnotations(List<TestAnnotationBuildItem> items, BuildProducer<BeanDefiningAnnotationBuildItem> producer) {
@@ -44,6 +57,37 @@ public class TestsAsBeansProcessor {
             builder.addBeanClass(item.getTestClassName());
         }
         producer.produce(builder.build());
+    }
+
+    @BuildStep(onlyIf = IsTest.class)
+    AnnotationsTransformerBuildItem vetoTestClassesNotMatchingTestProfile(Optional<TestProfileBuildItem> testProfile,
+            CombinedIndexBuildItem index) {
+        // Since we only need to register all test classes that belong to the "current" test profile
+        // we need to veto other test classes during the build
+        return new AnnotationsTransformerBuildItem(AnnotationTransformation
+                .forClasses()
+                .when(tc -> {
+                    ClassInfo maybeTestClass = tc.declaration().asClass();
+                    boolean veto = false;
+                    if (maybeTestClass.hasAnnotation(QUARKUS_TEST)
+                            || maybeTestClass.hasAnnotation(QUARKUS_INTEGRATION_TEST)
+                            || maybeTestClass.hasAnnotation(QUARKUS_MAIN_TEST)) {
+                        String testProfileClassName = testProfile.map(TestProfileBuildItem::getTestProfileClassName)
+                                .orElse(null);
+                        veto = !matchesProfile(maybeTestClass, testProfileClassName);
+                        if (veto && hasMatchingNestedTest(maybeTestClass, testProfileClassName, index.getComputingIndex())) {
+                            // FIXME the current @Nested tests support makes it possible to specify a different test profile
+                            // which is wrong and may cause troubles if a test class injects beans enabled/disabled in a specific profile
+                            // See https://github.com/quarkusio/quarkus/issues/45349
+                            LOG.warnf(
+                                    "Test class [%s] does not match the current test profile [%s] but cannot be vetoed because it declares a matching @Nested test",
+                                    maybeTestClass, testProfileClassName);
+                            veto = false;
+                        }
+                    }
+                    return veto;
+                })
+                .transform(tc -> tc.add(AnnotationInstance.builder(DotNames.VETOED).buildWithTarget(tc.declaration()))));
     }
 
     @BuildStep(onlyIf = IsTest.class)
@@ -89,8 +133,6 @@ public class TestsAsBeansProcessor {
         });
     }
 
-    private static final DotName QUARKUS_TEST_PROFILE = DotName.createSimple("io.quarkus.test.junit.QuarkusTestProfile");
-
     private static Set<DotName> initTestProfileHierarchy(Optional<TestProfileBuildItem> testProfile, IndexView index) {
         Set<DotName> ret = Set.of();
         if (testProfile.isPresent()) {
@@ -130,6 +172,32 @@ public class TestsAsBeansProcessor {
             } else {
                 superName = null;
             }
+        }
+        return false;
+    }
+
+    private boolean hasMatchingNestedTest(ClassInfo testClass, String testProfileClassName, IndexView index) {
+        for (DotName memberClassName : testClass.memberClasses()) {
+            ClassInfo memberClass = index.getClassByName(memberClassName);
+            if (memberClass != null && memberClass.hasDeclaredAnnotation(NESTED)) {
+                if (matchesProfile(memberClass, testProfileClassName)
+                        || hasMatchingNestedTest(memberClass, testProfileClassName, index)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private boolean matchesProfile(ClassInfo testClass, String testProfileClassName) {
+        AnnotationInstance testProfileAnnotation = testClass.declaredAnnotation(TEST_PROFILE);
+        if (testProfileClassName == null) {
+            // No test profile set - match test classes without @TestProfile
+            return testProfileAnnotation == null;
+        } else if (testProfileAnnotation != null) {
+            // Test profile set - match test classes with the same test profile
+            String annotationProfileClassName = testProfileAnnotation.value().asClass().name().toString();
+            return testProfileClassName.equals(annotationProfileClassName);
         }
         return false;
     }

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/SharedProfileTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/SharedProfileTestCase.java
@@ -2,6 +2,7 @@ package io.quarkus.it.main;
 
 import static org.hamcrest.Matchers.is;
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.util.Collections;
 import java.util.List;
@@ -18,6 +19,7 @@ import jakarta.inject.Inject;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
+import io.quarkus.arc.Arc;
 import io.quarkus.it.rest.ExternalService;
 import io.quarkus.it.rest.GreetingService;
 import io.quarkus.test.common.QuarkusTestResourceLifecycleManager;
@@ -68,6 +70,7 @@ public class SharedProfileTestCase {
     public void testProfileBeans() {
         assertEquals("Bonjour Foo", greetingService.greet("Foo"));
         assertEquals("profile", externalService.service());
+        assertTrue(Arc.container().select(SharedNormalTestCase.class).isUnsatisfied());
     }
 
     public static class MyProfile implements QuarkusTestProfile {

